server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java

```diff
@@ -230,46 +231,71 @@ private class GetSnapshotsOperation {
-                    allSnapshotInfos.stream().flatMap(Collection::stream),
-                    totalCount.get(),
-                    offset,
-                    size
-                );
-                final var snapshotInfos = sortedSnapshotsInRepos.snapshotInfos();
-                assert indices || snapshotInfos.stream().allMatch(snapshotInfo -> snapshotInfo.indices().isEmpty());
-                final int finalRemaining = sortedSnapshotsInRepos.remaining() + remaining.get();
-                return new GetSnapshotsResponse(
-                    snapshotInfos,
-                    failuresByRepository,
-                    finalRemaining > 0 ? sortBy.encodeAfterQueryParam(snapshotInfos.get(snapshotInfos.size() - 1)) : null,
-                    totalCount.get(),
-                    finalRemaining
-                );
-            }))) {
-                for (final RepositoryMetadata repository : repositories) {
-                    final String repoName = repository.name();
-                    if (skipRepository(repoName)) {
-                        // TODO we should still count the matching snapshots in totalCount
-                        continue;
-                    }
+            SubscribableListener
+
+                .<Void>newForked(repositoriesDoneListener -> {
+                    try (var listeners = new RefCountingListener(repositoriesDoneListener)) {
+                        for (final RepositoryMetadata repository : repositories) {
+                            final String repoName = repository.name();
+                            if (skipRepository(repoName)) {
+                                continue;
+                            }
 
-                    getSingleRepoSnapshotInfo(repoName, listeners.acquire((SnapshotsInRepo snapshotsInRepo) -> {
-                        allSnapshotInfos.add(snapshotsInRepo.snapshotInfos());
-                        remaining.addAndGet(snapshotsInRepo.remaining());
-                        totalCount.addAndGet(snapshotsInRepo.totalCount());
-                    }).delegateResponse((l, e) -> {
-                        if (isMultiRepoRequest && e instanceof ElasticsearchException elasticsearchException) {
-                            failuresByRepository.put(repoName, elasticsearchException);
-                            l.onResponse(SnapshotsInRepo.EMPTY);
-                        } else {
-                            l.onFailure(e);
+                            SubscribableListener
+
+                                .<RepositoryData>newForked(repositoryDataListener -> {
+                                    if (snapshotNamePredicate == SnapshotNamePredicate.MATCH_CURRENT_ONLY) {
+                                        repositoryDataListener.onResponse(null);
+                                    } else {
+                                        repositoriesService.repository(repoName).getRepositoryData(executor, repositoryDataListener);
```

Comment:
Previously we used `RepositoriesService#getRepositoryData` which completes its listener on some random threadpool (often the calling one, but not always, could even be the master thread). This fixes that.